// node-enterprise/build/src/logger/log.js
"use strict";
const inspect_1 = require("./inspect");
/**
 * @license Mit Licence 2015
 * @since 1.0.0
 * @name Log
 * @private
 * @constructor
 * @description
 * Log is private class used by logger to present logs to outputs
 */
class Log {
    constructor(message, data, level) {
        this.message = message;
        this.level = level;
        this.created = new Date().toISOString();
        this.data = data instanceof Error ? data.stack : data;
    }
    /**
     * @since 1.0.0
     * @function
     * @name Log#getName
     *
     * @description
     * Get log level name
     */
    getName() {
        return this.level.getName();
    }
    /**
     * @since 1.0.0
     * @function
     * @name Log#getLevel
     *
     * @description
     * Get log level name
     */
    getLevel() {
        return this.level.getLevel();
    }
    /**
     * @since 1.0.0
     * @function
     * @name Level#prettify
     *
     * @description
     * Prettify output and print it
     */
    prettify(prettify) {
        let log = inspect_1.inspect({
            created: this.created,
            data: this.data,
            level: this.getLevel(),
            message: this.message,
            type: this.getName()
        }, 5);
        if (!!prettify) {
            log = inspect_1.clean(log);
        }
        return log;
    }
    /**
     * @since 1.0.0
     * @function
     * @name Level#toString
     *
     * @description
     * Convert log to string so we can write it on file for example
     */
    toString() {
        return JSON.stringify(this.prettify(true));
    }
}
exports.Log = Log;
